lib/load/loadbalancer.go

```diff
@@ -0,0 +1,135 @@
+*/
+type Balancer struct {
+	hosts  []*host
+	dialer *net.Dialer
+	mu     sync.Mutex
+}
+
+/*
+NewLoadBalancer constructs a new least-connections load balancer to route
+requests to the slice of TCP addresses provided. Clients should construct a separate
+LoadBalancer for each upstream application they wish to load balance.
+
+An error is returned in any of the following scenarios:
+
+- The slice of addresses passed is empty
+- The slice of addresses passed is nil
+- The slice of addresses contains only nil entries
+
+Duplicate addresses are ignored. If two addresses passed share the same IP, zone and port they
+will be treated as a single host when performing load balancing.
+*/
+func NewLoadBalancer(addresses []*net.TCPAddr) (*Balancer, error) {
+	validateAddresses, err := validateAndRemoveDuplicateAddresses(addresses)
+	if err != nil {
+		return nil, err
+	}
+
+	hosts := make([]*host, 0, len(validateAddresses))
+	for _, address := range validateAddresses {
+		host := &host{address: address, connectionCount: &atomic.Counter{}}
+		hosts = append(hosts, host)
+	}
+
+	// I believe when both a context and a dialer have a timeout the shorter value
+	// is respected; this protects us from clients passing in a no-timeout context
+	// and our dial deadlocking when we can't connect to the upstream.
+	return &Balancer{hosts: hosts, dialer: &net.Dialer{Timeout: maxConnectionTimeout}, mu: sync.Mutex{}}, nil
+}
+
+/*
+HandleConnection takes a TCP connection, finds a suitable host to handle it,
+then connects to the host and streams data between the connection and host until
+both sides of the connection are closed.
+
+If the connection to the host fails, HandleConnection will write an error message
+to the incoming net.Conn and close it. It will also close the net.Conn if the communication
+with the host succeeds; callers of HandleConnection do not need to close the net.Conn
+they pass in.
+*/
+func (b *Balancer) HandleConnection(ctx context.Context, conn net.Conn) {
+	host := b.findHostWithLeastConnections()
+
+	host.connectionCount.Increment()
+	defer host.connectionCount.Decrement()
+
+	connectionToHost, err := b.dialer.DialContext(ctx, "tcp", host.address.String())
+
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			conn.Write([]byte(connectionToUpstreamTimedOutMessage))
+		default:
+			conn.Write([]byte(internalServerErrorMessage))
+		}
+		closeErr := conn.Close()
+		if closeErr != nil {
+			log.Printf("Error closing connection after dial failure. Dial error: %s, connection close error: %s", err.Error(), closeErr.Error())
+		}
+		return
+	}
+
+	var waitGroup sync.WaitGroup
+
+	waitGroup.Add(2)
+
+	go func() {
+		defer conn.Close()
+		io.Copy(conn, connectionToHost)
+		waitGroup.Done()
+	}()
+
+	go func() {
+		defer connectionToHost.Close()
+		io.Copy(connectionToHost, conn)
+		waitGroup.Done()
+	}()
+
+	waitGroup.Wait()
+}
+
+func (b *Balancer) findHostWithLeastConnections() *host {
+	host := b.hosts[0]
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
```

Comment:
Per [this discussion](https://github.com/KiaFarhang/tcp-load-balancer/pull/2#discussion_r875392544), mutex would prevent multiple requests triggering iterations over the list of hosts at once. I think the `host := b.hosts[0]` is safe because we don't support modifying the slice of hosts after construction, so every thread could read that first one at the same time - it won't change. What do you think?